flash-lso/src/amf3/read.rs

```diff
@@ -464,145 +473,225 @@ impl AMF3Decoder {
+                if i.len() < len * 4 {
+                    return Err(Err::Error(make_error(i, ErrorKind::TooLarge)));
+                }
+                let (i, fixed_length) = be_u8(i)?;
 
-            let (i, ints) = many_m_n(len, len, be_u32)(i)?;
+                let (i, ints) = many_m_n(len, len, be_u32)(i)?;
 
-            Ok((i, Value::VectorUInt(ints, fixed_length == 1)))
-        })
+                Ok((i, Value::VectorUInt(ints, fixed_length == 1)))
+            },
+        )
     }
 
     fn parse_element_vector_double<'a>(&mut self, i: &'a [u8]) -> AMFResult<'a, Rc<Value>> {
-        self.parse_reference_or_val(i, |_this, i, len| {
-            // There must be at least `len * 8` (f64 = 8 bytes) bytes to read this, this prevents OOM errors with v.large dicts
-            if i.len() < len * 8 {
-                return Err(Err::Error(make_error(i, ErrorKind::TooLarge)));
-            }
-            let (i, fixed_length) = be_u8(i)?;
+        self.parse_reference_or_val(
+            i,
+            |_| Value::VectorDouble(Vec::new(), false),
+            |_this, i, len, _| {
+                // There must be at least `len * 8` (f64 = 8 bytes) bytes to read this, this prevents OOM errors with v.large dicts
+                if i.len() < len * 8 {
+                    return Err(Err::Error(make_error(i, ErrorKind::TooLarge)));
+                }
+                let (i, fixed_length) = be_u8(i)?;
 
-            let (i, numbers) = many_m_n(len, len, be_f64)(i)?;
+                let (i, numbers) = many_m_n(len, len, be_f64)(i)?;
 
-            Ok((i, Value::VectorDouble(numbers, fixed_length == 1)))
-        })
+                Ok((i, Value::VectorDouble(numbers, fixed_length == 1)))
+            },
+        )
     }
 
     fn parse_element_object_vector<'a>(&mut self, i: &'a [u8]) -> AMFResult<'a, Rc<Value>> {
-        self.parse_reference_or_val(i, |this, i, len| {
-            let (i, fixed_length) = be_u8(i)?;
+        self.parse_reference_or_val(
+            i,
+            |this| {
+                this.object_id += 1;
+
+                Value::VectorObject(ObjectId(this.object_id), Vec::new(), "".to_string(), false)
+            },
+            |this, i, len, ofi| {
+                let (i, fixed_length) = be_u8(i)?;
+
+                let (i, object_type_name) = this.parse_string(i)?;
 
-            let (i, object_type_name) = this.parse_string(i)?;
+                let (i, elems) = many_m_n(len, len, |i| this.parse_single_element(i))(i)?;
 
-            let (i, elems) = many_m_n(len, len, |i| this.parse_single_element(i))(i)?;
 
-            Ok((
-                i,
-                Value::VectorObject(elems, object_type_name, fixed_length == 1),
-            ))
-        })
+                let id = if let Value::VectorObject(id, _, _, _) =
+                    this.object_reference_table.get(ofi).unwrap().as_ref()
+                {
+                    id
+                } else {
+                    unreachable!()
```

Comment:
nit: A message would be appreciated here, for example like at `Value::ECMAArray`.

Reply:
Done